TCP netstat utility, new tcp_sources directory v2

htibosch opened the pull request:
<!--- Title -->

Description
-----------

( this PR replaces #58, which got messed up while rebasing, sorry for that. ).

A few things happened in this PR:

1) created 2 new directories where non-essential but useful sources can be stored:
~~~
    tools/tcp_sources
    tools/tcp_sources/include
~~~

2) Created a function that retrieves information about sockets, open ports, and some basic I/O statistics:
~~~
    tools/tcp_sources/tcp_netstat.c
    tools/tcp_sources/include/tcp_netstat.h
~~~

3) move some sources code from the tools directory into the new directories:
~~~
    tools/tcp_sources/tcp_mem_stats.c
    tools/tcp_sources/tcp_dump_packets.c
    tools/tcp_sources/tcp_mem_stats.md
    tools/tcp_sources/tcp_dump_packets.md
    tools/tcp_sources/include/tcp_mem_stats.h
    tools/tcp_sources/include/tcp_dump_packets.h
~~~

4) Removed the last few tabs, mostly found in comments.

5) Removed the explicit `#include` of `FreeRTOSIPConfigDefaults.h` in some c modules.

Test Steps
-----------
Make sure that your FreeRTOSIPConfig.h includes tools/tcp_sources/include/tcp_netstat.h:

    `@include "tcp_netstat.h"`

Tun the application for a while, bot UDP and TCP sockets. And then execute the following code:
~~~c
    MetricsType_t xMetrics;
    vGetMetrics( &( xMetrics ) );
    vShowMetrics ( &( xMetrics ) );
~~~
The function `vShowMetrics()` will print a list of all information, by using the `FreeRTOS_printf()` macro.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

AniruddhaKanhere commented:
Hello @htibosch,

Thanks for creating this PR. As you know, the spell check uses American English and thus is flagging the use of `synchronisation` and `synchronised`. Can you add these words (alphabetically and in NON capital case) to the file `.github/lexicon.txt` so that they can be ignored by the spell check?

Thanks,
Aniruddha

AniruddhaKanhere commented on tools/tcp_sources/include/tcp_netstat.h:
```diff
@@ -0,0 +1,118 @@
+} MetricsType_t;
+
+extern BaseType_t vGetMetrics( MetricsType_t * pxMetrics );
+extern void vShowMetrics( MetricsType_t * pxMetrics );
```
```suggestion
extern void vShowMetrics( const MetricsType_t * pxMetrics );
```
Also see suggestion in the corresponding c file.

htibosch replied:
You're right, added a const.

AniruddhaKanhere commented on tools/tcp_sources/tcp_netstat.c:
```diff
@@ -0,0 +1,194 @@
+    memset( pxMetrics, 0, sizeof *pxMetrics );
+
+    pxMetrics->xInput = xInputCounters;
+    pxMetrics->XOutput = xOutputCOunters;
```
```suggestion
    pxMetrics->XOutput = xOutputCounters;
```

htibosch replied:
Not only a hanging shift, also copy/pasted the typo :-)

htibosch commented on tools/tcp_sources/include/tcp_netstat.h:
```diff
@@ -0,0 +1,118 @@
+    size_t uxPacketCount;
+} IOCounters_t;
+
+extern IOCounters_t xInputCounters, xOutputCOunters;
```
A hanging shift button, repaired

htibosch commented on tools/tcp_sources/include/tcp_netstat.h:
```diff
@@ -0,0 +1,118 @@
+    xInputCounters.uxPacketCount++;
+
+#define iptraceNETWORK_INTERFACE_OUTPUT( uxDataLength, pucEthernetBuffer ) \
+    xOutputCOunters.uxByteCount += uxDataLength;                           \
+    xOutputCOunters.uxPacketCount++;
```
Same as above, corrected.

htibosch commented on tools/tcp_sources/tcp_netstat.c:
```diff
@@ -0,0 +1,194 @@
+    extern List_t xBoundTCPSocketsList;
+#endif /* ipconfigUSE_TCP == 1 */
+
+IOCounters_t xInputCounters, xOutputCOunters;
```
Same again...

htibosch commented on tools/tcp_sources/tcp_netstat.c:
```diff
@@ -0,0 +1,194 @@
+    return xResult;
+}
+
+void vShowMetrics( MetricsType_t * pxMetrics )
```
Done so

htibosch commented on tools/tcp_sources/tcp_netstat.md:
```diff
@@ -0,0 +1,19 @@
+
+    `BaseType_t vGetMetrics( MetricsType_t * pxMetrics )`
+
+It collect information: all port numbers in use, all UDP sockets, all TCP sockets and their connections.
```
yes, well seen.

htibosch reviewed (commented):
Thanks for you review.
I hope that we can also agree on the changed directory structure.
Hein

htibosch commented:
About the last few changes:

Renamed some macro's: e.g. "MAX_UDP_PORTS" is too generic. I gave these macro's the prefix nstat as in "nstatMAX_UDP_PORTS".

We found "tcp_sources" confusing, because there is already a TCP source directory at the root of the repo.
Therefore, I renamed "tools/tcp_sources" to "tools/tcp_utilities".
